### PR TITLE
Add AbstractFormFieldDecorator

### PR DESCRIPTION
**Pull request by Hanashi**
See https://community.woltlab.com/thread/292040-formbuilder-feld-nicht-in-data-eintragen/

### CODE REVIEW

File: wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+     * @inheritDoc
+     */
+    public static function create($id)
+    {
+        throw new \BadMethodCallException('This method is available on a decorator.');
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php:458]:
```suggestion
    /**
     * @throws \BadMethodCallException always, as static methods are not defined on decorators.
     */
```

File: wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+    public static function validateAttribute($name)
+    {
+        throw new \BadMethodCallException('This method is available on a decorator.');
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php:466]:
```suggestion
    /**
     * @throws \BadMethodCallException always, as static methods are not defined on decorators.
     */
```

File: wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+    public static function validateClass($class)
+    {
+        throw new \BadMethodCallException('This method is available on a decorator.');
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php:474]:
```suggestion
    /**
     * @throws \BadMethodCallException always, as static methods are not defined on decorators.
     */
```

File: wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+namespace wcf\system\form\builder\field;
+
+use wcf\data\IStorableObject;
+use wcf\system\form\builder\field\dependency\IFormFieldDependency;
+use wcf\system\form\builder\field\validation\IFormFieldValidationError;
+use wcf\system\form\builder\field\validation\IFormFieldValidator;
+use wcf\system\form\builder\IFormParentNode;
+
+/**
+ * Decorator for form field objects.
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php:12]:
```suggestion
 * Default implementation of a decorator for an `IFormField` that forwards all methods defined
 * in the interface to the implementation of the decorated field.
```

File: wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+{
+    /**
+     * decorated field
+     *
+     * @var IFormField
+     */
+    protected $field;
+
+    /**
+     * Creates a decorator object for form fields
```
Comment (TimWolla) [inline wcfsetup/install/files/lib/system/form/builder/field/AbstractFormFieldDecorator.class.php:30]: This is pretty tautological as well. I can't really think of a better comment either. Maybe just completely remove the PHPDoc for `__construct()`.

**Review by TimWolla — CHANGES_REQUESTED** [review]

**Review by TimWolla — APPROVED** [review]
Thanks!

**Comment by mutec** [comment]
@TimWolla the `@since`-doc is incorrect.

**Comment by TimWolla** [comment]
> @TimWolla the `@since`-doc is incorrect.

Yes, technically it's 5.4.something. However running an outdated version of a specific branch is unsupported in practice.

**Comment by mutec** [comment]
However it's inconsistent since WoltLab announced to use SemVer strictly. ;)

**Comment by dtdesign** [comment]
> However it's inconsistent since WoltLab announced to use SemVer strictly. ;)

That is not what SemVer means. It only guarantees the compatibility in one direction, adding self-contained functions that do not alter the API does not violate it.

**Pull request merged**